core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala

```diff
@@ -171,7 +175,7 @@ class HistoryServer(
    * @return List of all known applications.
    */
   def getApplicationList(): Iterable[ApplicationHistoryInfo] = {
-    provider.getListing()
+    provider.getListing().take(maxApplications)
```

Comment:
I think you cannot add ".take(maxApplications)" here since it may affect other places to get the list of all application history. The only problem is in rest endpoint "api/v1/applications".